README.md
# babylon_dart

Bindings for the [BabylonJS](https://www.babylonjs.com/) 4.2 web rendering engine.

## Status Quo

The bindings are created using a self-written generator ([babylon_dart_gen](https://github.com/blimster/babylon_dart_gen)). This is because the generator [js_facade_gen](https://github.com/dart-lang/js_facade_gen) from the Dart project runs into multiple exceptions when trying to generate bindings from the `babylon.d.ts` file.

`babylon_dart_gen` has a configuration to define what should be included or excluded. At the moment, only a subset is included. The more features are implemented in the generator the more parts of the `babylon.d.ts` file can be converted.

Work is in progress. Tell me what is important for you and I will try to include it in a new release of `babylon`. Or even better, create a pull request in the [babylon_dart_gen](https://github.com/blimster/babylon_dart_gen) project.

## Documentation

Currently, no docs are generated in the bindings. For help look at the official [docs](https://doc.babylonjs.com/).

## Example

In your `index.html` add links to the BabylonJS implementation:

```html
<script src="/packages/babylon_dart/assets/babylon.js"></script>
<script src="/packages/babylon_dart/assets/babylonjs.loaders.min.js"></script>
<script src="/packages/babylon_dart/assets/babylonjs.gui.min.js"></script>
<script src="/packages/babylon_dart/assets/babylonjs.serializers.min.js"></script>
<script src="/packages/babylon_dart/assets/babylonjs.inspector.bundle.js"></script>
<script src="/packages/babylon_dart/assets/cannon.js"></script>
```

Here is a little example rendering 2 meshes with a little physics in it:

```dart
import 'dart:html';

import 'package:babylon_dart/babylon_dart.dart';
import 'package:js/js.dart';

void main() {
  // query the canvas element
  final canvas = querySelector('#render-canvas') as CanvasElement;

  // create the engine
  final engine = Engine(canvas, true);

  // create the scene andd configure it
  final scene = Scene(engine);
  scene.clearColor = Color4(0.2, 0.2, 0.2, 1);

  // create a camera that can be controlled by mouse or keyboard
  final camera = ArcRotateCamera('camera', 1, 1, 20, Vector3(0, 0, 0), scene);
  camera.attachControl();

  // a global light
  final hemiLight = HemisphericLight("hemiLight", Vector3(0, 1, 0), scene);
  hemiLight.diffuse = Color3(0.5, 0.5, 0.5);

  // a point light for shadows
  final pointLight = PointLight("dirL", Vector3(0, 10, 0), scene);
  pointLight.diffuse = Color3(0.5, 0.5, 0.5);

  // create material and mesh for a red sphere
  final sphereMaterial = StandardMaterial("sphereMat", scene);
  sphereMaterial.diffuseColor = Color3.Red();

  final sphere = MeshBuilder.CreateSphere('sphere', MeshBuilderCreateSphereOptions(segments: 16), scene);
  sphere.material = sphereMaterial;
  sphere.position = Vector3(0, 5, 0);

  // create a textured ground
  final groundTexture = Texture.args('dart-logo.jpeg', scene, true, false, 1);
  groundTexture.uScale = 1;
  groundTexture.vScale = 1;

  final groundMaterial = StandardMaterial("groundMat", scene);
  groundMaterial.diffuseColor = Color3.White();
  groundMaterial.diffuseTexture = groundTexture;
  groundMaterial.specularColor = Color3(0.1, 0.1, 0.1);

  final ground = MeshBuilder.CreateGround('ground', MeshBuilderCreateGroundOptions(height: 10, width: 10), scene);
  ground.material = groundMaterial;
  ground.receiveShadows = true;

  // enable shadows
  final shadowGenerator = ShadowGenerator(512, pointLight);
  shadowGenerator.getShadowMap().renderList.add(sphere);

  // enabled physics
  scene.enablePhysics(Vector3(0, -9.81, 0), CannonJSPlugin());

  ground.physicsImpostor = PhysicsImpostor(
    ground as IPhysicsEnabledObject,
    PhysicsImpostor.HeightmapImpostor,
    PhysicsImpostorParameters(restitution: 4),
    scene,
  );

  sphere.physicsImpostor = PhysicsImpostor(
    sphere as IPhysicsEnabledObject,
    PhysicsImpostor.SphereImpostor,
    PhysicsImpostorParameters(mass: 1),
    scene,
  );

  // focus canvas for keyboard controls
  canvas.focus();

  // the render loop
  engine.runRenderLoop(allowInterop(() {
    scene.render();
  }));

  // react to changes of the size of the canvas
  window.onResize.listen((event) {
    engine.resize();
  });
}
```


CHANGELOG.md
# Changelog

## v1.1.1

- Updated to BabylonJS 4.2.1 (see https://forum.babylonjs.com/t/broken-shadows-after-recent-chrome-update-in-4-2/26479)

## v1.1.0

- BREAKING CHANGE: Requires Dart SDK 2.14.0
- Added binding for Angle and Gizmos

## v1.0.0

- Stable null safety release.

## v1.0.0-nullsafety.2

- Added bindings for `RecastJSPlugin`.

## v1.0.0-nullsafety.1

- BREAKING CHANGE: The constructor with parameters of class Engine is now a named constructor 'args'
- Added bindings for NullEngine.

## v1.0.0-nullsafety.0

- First null safe version.

## v0.8.0

- BREAKING CHANGE: The constructor with parameters of class Engine is now a named constructor 'args'
- Added bindings for NullEngine.

## v0.7.4

- Added biindings for physically based rendering.

## v0.7.3

- Added classes GLTF2Export, IExportOptions and GLTFData.

## v0.7.2

- Added class CSG.

## v0.7.1

- Added basics for Solid Particle System (simple example with some animation works).

## v0.7.0

- BREAKING CHANGE: IInspectorOptions is now handled as object literal.

## v0.6.1

- Added scripts to run BabylonJS inspector.

## v0.6.0+1

- Little changes in README.md and added some comments to the examplle.

## v0.6.0

- BREAKING CHANGE: BabylonJS 4.2

## v0.5.1

- Added binding for DirectionalLight

## v0.5.0

- More bindings (including classes for TypeScript enums)
- BREAKING CHANGE: Enusure a no arg contructor for extended classes. If such a class have already a constructor with args, it will be named 'args'.

## v0.4.0

- Much more bindings
- BREAKING CHANGE: 2.10.0 as minimum SDK constaint

## v0.3.0

- Much more bindings
- All bindings are generated now
- Bugifxes
- BREAKING CHANGES: because all code is generated now and bugs were fixed, there are breaking changes in the API.

## v0.2.0

-   Added support for optional parameters
-   Added support for type parameters on classes
-   Added `Promise` type
-   Added support for abstract classes and methods

## v0.1.1+1

-   Fixed spelling in README.md

## v0.1.1

-   added ShadowLight

## v0.1.0

-   Initial version


example/main.dart
import 'dart:html';

import 'package:babylon_dart/babylon_dart.dart';
import 'package:js/js.dart';

void main() {
  // query the canvas element
  final canvas = querySelector('#render-canvas') as CanvasElement;

  // create the engine
  final engine = Engine(canvas, true);

  // create the scene andd configure it
  final scene = Scene(engine);
  scene.clearColor = Color4(0.2, 0.2, 0.2, 1);

  // create a camera that can be controlled by mouse or keyboard
  final camera = ArcRotateCamera('camera', 1, 1, 20, Vector3(0, 0, 0), scene);
  camera.attachControl();

  // a global light
  final hemiLight = HemisphericLight("hemiLight", Vector3(0, 1, 0), scene);
  hemiLight.diffuse = Color3(0.5, 0.5, 0.5);

  // a point light for shadows
  final pointLight = PointLight("dirL", Vector3(0, 10, 0), scene);
  pointLight.diffuse = Color3(0.5, 0.5, 0.5);

  // create material and mesh for a red sphere
  final sphereMaterial = StandardMaterial("sphereMat", scene);
  sphereMaterial.diffuseColor = Color3.Red();

  final sphere = MeshBuilder.CreateSphere('sphere', MeshBuilderCreateSphereOptions(segments: 16), scene);
  sphere.material = sphereMaterial;
  sphere.position = Vector3(0, 5, 0);

  // create a textured ground
  final groundTexture = Texture.args('dart-logo.jpeg', scene, true, false, 1);
  groundTexture.uScale = 1;
  groundTexture.vScale = 1;

  final groundMaterial = StandardMaterial("groundMat", scene);
  groundMaterial.diffuseColor = Color3.White();
  groundMaterial.diffuseTexture = groundTexture;
  groundMaterial.specularColor = Color3(0.1, 0.1, 0.1);

  final ground = MeshBuilder.CreateGround('ground', MeshBuilderCreateGroundOptions(height: 10, width: 10), scene);
  ground.material = groundMaterial;
  ground.receiveShadows = true;

  // enable shadows
  final shadowGenerator = ShadowGenerator(512, pointLight);
  shadowGenerator.getShadowMap().renderList.add(sphere);

  // enabled physics
  scene.enablePhysics(Vector3(0, -9.81, 0), CannonJSPlugin());

  ground.physicsImpostor = PhysicsImpostor(
    ground as IPhysicsEnabledObject,
    PhysicsImpostor.HeightmapImpostor,
    PhysicsImpostorParameters(restitution: 4),
    scene,
  );

  sphere.physicsImpostor = PhysicsImpostor(
    sphere as IPhysicsEnabledObject,
    PhysicsImpostor.SphereImpostor,
    PhysicsImpostorParameters(mass: 1),
    scene,
  );

  // focus canvas for keyboard controls
  canvas.focus();

  // the render loop
  engine.runRenderLoop(allowInterop(() {
    scene.render();
  }));

  // react to changes of the size of the canvas
  window.onResize.listen((event) {
    engine.resize();
  });
}
